Updates needed to collect EFS (and later S3) usage data

ahaywardtvuk opened the pull request:
This adds the concept of a consumption rate sample to the accounting service. This is a point-in-time sample of how quickly a billable resource is being consumed from which we can estimate the amount actually consumed over some period.

This is intended for storage, where the billable resource is an amount measured in gigabyte-seconds. If we measure your workspace store at 8GB then this means you're using 8GB-s per second.

This change includes:

* The Messager which ingests and records consumption rate samples.
* Code in this Messager to generate BillingEvents from the rate samples periodically.
* A change to use datetimes with a timestamp attached everywhere so there's no ambiguity about which zone they're in. PostgreSQL is now using TIMESTAMPTZ columns (which doesn't actually record a timezone, just takes it into account when you give it one).
* When creating stub BillingItems for SKUs we've observed but not got in the DB, be more careful not to create two if two instances of the service are running.

ahaywardtvuk committed:
Add a method to estimate resource use from resource use-rate samples.

ahaywardtvuk committed:
Must use exactly the Python package name in log_component_version

ahaywardtvuk committed:
Add code to generate estimated BillingEvents from consumption rate data at the appropriate time.
This includes the consumption rate messager which will record consumption rates.

ahaywardtvuk committed:
Switch to always specifying a timezone in our datetimes.
This will require recreating the DB.

ahaywardtvuk committed:
Set up the sample rate ingester to receive Pulsar messages.

ahaywardtvuk committed:
find_latest_billing_event should return a BillingEvent or None, not a tuple.

ahaywardtvuk committed:
Use required branch of eodhp-utils

ahaywardtvuk committed:
Update dependencies

ahaywardtvuk committed:
Try harder not to insert duplicate inferred BillingItems when we have two ingesters running.

ahaywardtvuk committed:
Get the tests working with sqlite as well as PostgreSQL

ahaywardtvuk committed:
eodhp-utils with required features now tagged - depend on v0.1.8 instead of branch.

ahaywardtvuk committed:
Be more defensive about getting a datetime with a timezone set.

james-hinton commented on accounting_service/models.py:
```diff
@@ -299,3 +374,218 @@ def __repr__(self):
+        return session.execute(query).scalars().all()
+
+    @classmethod
+    def calculate_consumption_for_interval(
```
Is there a scenario where we would have a single-sample, would that cause an `IndexError`?

ahaywardtvuk replied:
I've added a check to handle that case by returning None and making the messager treat it as an estimate of zero consumption.

We assume zero consumption up to our first ever sample and zero consumption after our last ever sample, so this treats this as consumption for a zero-length time window.

ahaywardtvuk committed:
Handle the case of having only one consumption rate sample when trying to estimate
consumption in an interval (answering zero consumption).

ahaywardtvuk committed:
Record consumption of zero when calculate_consumption_for_interval can't calculate an estimate.
This can only happen if we have only a single sample and this sample will be after the
window being estimated for, so we assume the resource didn't exist yet.
